src/runtime/virtcontainers/kata_agent.go

```diff
@@ -379,6 +381,43 @@ func (k *kataAgent) internalConfigure(h hypervisor, id string, config interface{
 	return nil
 }
 
+func setupSandboxBindMounts(sandbox *Sandbox) error {
+	if len(sandbox.config.SandboxBindMounts) == 0 {
+		return nil
+	}
+
+	// Create subdirectory in host shared path for sandbox mounts
+	sandboxMountDir := filepath.Join(getMountPath(sandbox.id), sandboxMountsDir)
+	if err := os.MkdirAll(sandboxMountDir, DirMode); err != nil {
+		return fmt.Errorf("Creating sandbox shared mount directory: %v: %w", sandboxMountDir, err)
+	}
+
+	for _, m := range sandbox.config.SandboxBindMounts {
+		mountDest := filepath.Join(sandboxMountDir, filepath.Base(m))
+		// bind-mount each sandbox mount that's defined into the sandbox mounts dir
+		if err := bindMount(context.Background(), m, mountDest, true, "private"); err != nil {
```

Comment:
In testing, I'm seeing issues, in particular when the source of the bind mount is a path, rather than a regular file. I am only seeing the file created from inside the guest, but on the host, I see the entire directory bind mounted seemingly correctly:

ex: 

from the host shared dir:
```
[root@fedora repro-test]# tree  **/sandbox-mounts/
mounts/sandbox-mounts/
└── certs
    └── hi.pem
shared/sandbox-mounts/
└── certs
    └── hi.pem

```

from inside the guest (debug-console):
```
bash-4.2# pwd
pwd
/run/kata-containers/shared/containers/sandbox-mounts
bash-4.2# ls
ls
certs
bash-4.2# ls -lR
ls -lR
.:
total 0
drwxr-xr-x 2 root root 40 Mar  2 01:32 certs

./certs:
total 0
```

@lifupan @bergwolf  - anything obvious I'm missing? I expect this should be the same logic that is used for regular container mounts, ie: https://github.com/kata-containers/kata-containers/blob/90a18e228b064077d9ec0d4849034804486c9e37/src/runtime/virtcontainers/container.go#L474-L477

Reply:
Is it something similar to https://github.com/kata-containers/kata-containers/commit/a958eaa8d3211ae465c193923de61d3162dc37f0. MS_REMOUNT does not propagate to other subtrees.

Reply:
It wasn't that, as I do see the appropriate permissions and directory structure in both the mounts and shared directories on the host.

What it ended up being was that if I create the bind mount into the shared path _before_ starting virtiofsd/vmm, the contents of the directory would never propagate. If I create it right after starting the VMM, the entire directory is propagated through as expected.

I was surprised by this; does this match your expectations?  @dagrh @stefanha does this match your expectations?

Reply:
@egernst Some time ago @xanclic fixed a big in virtiofsd where sub-mounts did not propagate at runtime and may have suggestions. I think there are two possibilities: 1. You have hit a bug in virtiofsd that should be fixed, 2. You are using an old virtiofsd and the bug has already been fixed.

Reply:
I think this is the case of running older virtiofsd. I tested this with latest upstream qemu/virtiofsd and I can see files inside directories which are bind mounted under shared directory. If "announce_submounts" is enabled, then I also see the mount point inside the VM.

Reply:
IOW, I tested this.

mkdir /tmp/source /mnt/shared/dest-mnt/
touch /tmp/source/foo.txt
mount --bind /tmp/source /mnt/shared/dest-mnt/

Now run virtiofsd and VM and export /mnt/shared directory into guest.

Mount virtiofs inside guest.
mount -t virtiofs myfs /mnt/virtiofs/

I can now see /mnt/virtiofs/dest-mnt/foo.txt

Reply:
ACK! I thought we back ported all changes to 5.0 -- let me test with latest on QEMU master?

Reply:
I think following commit from @XanClic fixed the issue. And looks like it should be part of 5.1 qemu.

commit ace0829c0d08f0e5f1451e402e94495bc2166772
Author: Max Reitz <mreitz@redhat.com>
Date:   Fri Apr 24 15:35:16 2020 +0200

    virtiofsd: Show submounts
    
    Currently, setup_mounts() bind-mounts the shared directory without
    MS_REC.  This makes all submounts disappear.
    
    Pass MS_REC so that the guest can see submounts again.

Reply:
Even when building with latest, I'm still seeing the failure. I'm working to reproduce without Kata involved, but am getting basic failures when launching. First, to start virtiofsd:
```
virtiofsd --socket-path=/tmp/vhostqemu2 -o source=/tmp/virtio-fs-src -o cache=none -o announce_submounts -d
```

Then start QEMU:

```
qemu-system-x86_64 -machine pc,accel=kvm,kernel_irqchip,nvdimm -m 2048M,slots=10,maxmem=8949M \
        -device nvdimm,id=nv0,memdev=mem0 \
        -object memory-backend-file,id=mem0,mem-path=/usr/share/kata-containers/kata-containers.img,size=536870912 \
        -global kvm-pit.lost_tick_policy=discard -nographic -vga none --no-reboot \
        -kernel /usr/share/kata-containers/vmlinuz-5.4.77 \
        -append "tsc=reliable no_timer_check rcupdate.rcu_expedited=1 i8042.direct=1 i8042.dumbkbd=1 i8042.nopnp=1 i8042.noaux=1 noreplace-smp reboot=k console=ttyS0 cryptomgr.notests net.ifnames=0 pci=lastbus=0 root=/dev/pmem0p1 rootflags=dax,data=order
ed,errors=remount-ro ro rootfstype=ext4 quiet systemd.show_status=false panic=1 nr_cpus=8 systemd.unit=kata-containers.target systemd.mask=systemd-networkd.service systemd.mask=systemd-networkd.socket agent.debug_console" -smp 1,cores=1,threads=1,sockets
=8,maxcpus=8 \
        -device pcie-root-port,id=root1,slot=0 \
        -chardev socket,id=char0,path=/tmp/vhostqemu2 \
        -device vhost-user-fs-pci,id=myvirtiofs,chardev=char0,tag=myfs,queue-size=1024
```

However, as soon as this starts, virtiofsd fails with:
```
[ID: 00000001] fv_panic: libvhost-user: Invalid vring_addr message
```

Reply:
@egernst -object memory-backend-file ...,share=on is needed for vhost-user. I recently sent a QEMU patch that prints a clearer error message when share=on is missing, so this will become easier to troubleshoot in the future.